hopperi/Main.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE PatternGuards  #-}
+module Main where
+
+import Control.Arrow (left)
+import Control.Lens
+import Control.Monad.STE
+import qualified Data.Text as T
+import qualified Data.Map as Map
+import qualified Data.Vector as V
+import Text.Parsec (parse)
+
+import Data.HopperException
+import System.Console.Haskeline
+
+import Hopper.Internal.Core.Literal
+import Hopper.Internal.LoweredCore.ANF (toAnf)
+import Hopper.Internal.LoweredCore.ClosureConversion (closureConvert)
+import Hopper.Internal.LoweredCore.ClosureConvertedANF
+import Hopper.Internal.LoweredCore.EvalClosureConvertedANF as Eval
+import Hopper.Internal.Runtime.Heap
+import Hopper.Internal.Runtime.HeapRef
+import Hopper.Parsing.Term (pExpr)
+import Hopper.Parsing.Concrete (concreteToAbstract)
+
+main :: IO ()
+main =
+  let green = "\ESC[32m\STX"
+      red = "\ESC[31m\STX"
+      blue = "\ESC[34m\STX"
+      end = "\ESC[0m\STX"
+      loop :: InputT IO ()
+      loop = do
+        minput <- getInputLine $ blue ++ "hopper> " ++ end
+        case minput of
+          Nothing -> return ()
+          Just input -> do
+            let evaled = hopperEval input
+            outputStrLn $ case evaled of
+              Left str -> red ++ str ++ end
+              Right str -> green ++ str ++ end
+            loop
+
+      settings = Settings {
+        complete=noCompletion,
+        historyFile=Nothing,
+        autoAddHistory=True
+      }
+  in runInputT settings $ withInterrupt loop
+
+hopperEval :: String -> Either String String
+hopperEval input = do
+  concrete <- left show $ parse pExpr "repl" input
+  let abstract = concreteToAbstract concrete
+
```

Comment:
for the current rep, are you checking that only single arity tuple arguments are provided?